src/sim/simulate.cc

```diff
@@ -293,22 +293,30 @@ doSimLoop(EventQueue *eventq)
         assert(curTick() <= eventq->nextTick() &&
                "event scheduled in the past");
 
-        if (mainQueue && async_event) {
+        if (async_event) {
             async_event = false;
```

Comment:
There is a risk that you will be losing IO and stat events after this patch because async_event is cleared even if there are pending IO or stat signals. This notification interface is a bit of a mess and should probably be replaced with an std::atomic_uint32_t bit field instead of separate variables.

Reply:
I'm not totally sure how to fix this as I don't quite understand the situation in which it happens. Could you provide a theoretical trace?

On the other hand... is it better or worse to potentially lose a stat signal (which, btw, I don't think is a commonly used interface) or where ctrl-c simply doesn't do anything for seconds to minutes at a time?

I'm not going to have time to continue working on this beyond small updates. Our choices are (1) keep the current broken behavior or (2) go with something like this change. The best option, someone else picking this up and converting to some thread-safe implementation, seems quite unlikely in the short term. 

Do you think this change, on the whole, improves gem5, makes it worse, or makes no difference? If it's the latter two, then I'll drop the change. If it's the former, we should commit this and make a bigger change later.

Reply:
I'm not so concerned about the missing a stat signal, I'm more concerned about SIGIO. This change introduces a large the window where things can go wrong. Focusing on the SIGIO case, this chain of events would cause SIGIO to get lost:

1. Start processing non-main queue
2. SIGIO handler sets `async_event` and `async_io`.
3. `doSimLoop` clears `async_event` but doesn't service the SIGIO since it isn't in the main event loop.

After this sequence of events, `async_event` is false and `async_io` is true. Because of this, the SIGIO won't be processed until another asynchronous event happens.

Looking at the change again, I think it is also fundamentally broken when performing a multi-threaded simulation. The problem there is that each thread will end up posting its own exit event.

While I think it makes sense to think about this problem and maybe fix it in the future, I think this change is causing more trouble than it is worth. A user can still quit gem5 while it is performing initialisation (e.g., not just when using a private event queue) by sending a SIGQUIT. You can do this from the terminal with ctrl-\.